src/include/btree_cmp_inline.h

```diff
@@ -166,6 +306,145 @@ __wt_compare_bounds(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_ITEM *key, u
     return (0);
 }
 
+#ifdef HAVE_X86INTRIN_H
+/*
+ * __lex_compare_skip_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes, skipping leading bytes.
+ *     Returns: < 0 if user_item is lexicographically < tree_item = 0 if user_item is
+ *     lexicographically = tree_item > 0 if user_item is lexicographically > tree_item We use the
+ *     names "user" and "tree" so it's clear in the btree code which the application is looking at
+ *     when we call its comparison function.
+ */
+static WT_INLINE int
+__lex_compare_skip_gt_16(
+  const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp, size_t *matchp)
+{
+    __m128i res_eq, t, u;
+    uint64_t t64, u64;
+    int32_t eq_bits;
+    size_t match, final_match;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+
+    match = *matchp;
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp + match, treep = tstartp + match; uendp - userp > WT_VECTOR_SIZE;
```

Comment:
x86 generates [better code](https://www.godbolt.org/z/bGsd5v956) for an index-based loop rather than a pointer arithmetic based one. Compare with [your original](https://www.godbolt.org/z/zYK3358o3). This also has the other optimizations I mentioned below so you can see their impact as well.